### PR TITLE
Feature/evs 640

### PR DESCRIPTION
**Pull request by MounirKaroudEV**

### CODE REVIEW

File: simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/SetVariableMonitoringRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    @Override
+    public void handle(String callId, SetVariableMonitoringRequest request) {
+        List<SetMonitoringResult> results = new ArrayList<>();
+
+        for (SetMonitoringDatum data : request.getSetMonitoringData()) {
+            SetMonitoringResult monitoringResult = buildResponse(data);
+
+            String componentName = data.getComponent().getName().toString();
+            Optional<StationComponent> component = stationComponentsHolder.getComponent(new CiString.CiString50(componentName));
```
Comment (akoshterek) [inline simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/SetVariableMonitoringRequestHandler.java]: Such usage of Optional seems to be awkward a bit.
All execution paths return some object to be added into results list so `component.map().orElseGet()` seems to be natural

File: simulator-core/src/test/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/SetVariableMonitoringRequestHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    private SetMonitoringDatum createdMonitoringDatum(String componentName, String variableName) {
+        return new SetMonitoringDatum()
+                .withId(1)
+                .withType(SetMonitoringDatum.Type.LOWER_THRESHOLD)
+                .withSeverity(0)
+                .withComponent(new Component().withName(new CiString.CiString50(componentName)))
+                .withVariable(new Variable().withName(new CiString.CiString50(variableName)));
+    }
+
+}
```
Comment (akoshterek) [inline simulator-core/src/test/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/SetVariableMonitoringRequestHandlerTest.java]: empty line

File: simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponent.java [hunk 3]
```diff
@@ -59,6 +57,15 @@ public GetVariableResult getVariable(GetVariableDatum getVariableDatum) {
         return accessor.get(new AttributePath(component, variable, attributeType));
     }
 
+    /**
+     *  Retrieves the name of the variables for this specific component.
+     *
+     * @return Set with name of variables
+     */
+    public Set<String> getVariableNames() {
```
Comment (akoshterek) [inline simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponent.java]: do we support variable instances like var["blah"]?

File: simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/ClearVariableMonitoringRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+     *
+     * @param callId identity of the message
+     * @param request incoming request from the server
+     */
+    @Override
+    public void handle(String callId, ClearVariableMonitoringRequest request) {
+        List<ClearMonitoringResult> results = new ArrayList<>();
+        for (int id : request.getId()) {
+            ClearMonitoringResult result = new ClearMonitoringResult().withId(id);
+            if (stationComponentsHolder.clearMonitor(id)) {
```
Comment (osigida) [inline simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/ClearVariableMonitoringRequestHandler.java]: results.add(result.withStatus(stationComponentsHolder.clearMonitor(id) ? ClearMonitoringResult.Status.ACCEPTED) :ClearMonitoringResult.Status.NOT_FOUND) );

Comment (osigida) [reply simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/ClearVariableMonitoringRequestHandler.java]: `stationComponentsHolder.clearMonitor(id) ? ClearMonitoringResult.Status.ACCEPTED) : ClearMonitoringResult.Status.NOT_FOUND)` could be fetched to a method

File: simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponent.java (hunk 3 above)
Comment (MounirKaroudEV) [reply simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponent.java]: What do you mean?

**Comment by akoshterek** [comment]
Check Variable structure in spec , it has instance field

File: simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponent.java (hunk 3 above)
Comment (MounirKaroudEV) [reply simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponent.java]: Ok i saw it. I don't think it is supported right now, ad i did not see anything related to instances in the code

**Comment by akoshterek** [comment]
OK

**Comment by MounirKaroudEV** [comment]
Can you please review again now that i have added support also for N02?
Thanks

File: simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponentsHolder.java [hunk 5]
```diff
@@ -56,12 +58,43 @@ public StationComponentsHolder(Station station, StationStore stationStore) {
...
 
-    public void monitorComponent(int monitorId, String componentName, String variableName) {
+    public void monitorComponent(int monitorId, ComponentVariable componentVariable, SetMonitoringDatum datum) {
         MonitoredComponent monitored = monitoredComponents.getOrDefault(monitorId, new MonitoredComponent());
-        monitored.addMonitoredComponent(componentName, variableName);
+        monitored.addMonitoredComponent(componentVariable, datum);
         monitoredComponents.put(monitorId, monitored);
     }
 
+    public Map<ComponentVariable, List<SetMonitoringDatum>> getAllMonitoredComponents() {
```
Comment (akoshterek) [inline simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/component/StationComponentsHolder.java]: add some javadoc here please and for the next method as well please

File: simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/GetMonitoringReportRequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        if (monitoredComponents.isEmpty() || monitoredComponents.values().stream().allMatch(List::isEmpty)) {
+            stationMessageSender.sendCallResult(callId, new GetMonitoringReportResponse().withStatus(GetMonitoringReportResponse.Status.REJECTED));
+        } else {
+            stationMessageSender.sendCallResult(callId, new GetMonitoringReportResponse().withStatus(GetMonitoringReportResponse.Status.ACCEPTED));
+            stationMessageSender.sendNotifyMonitoringReport(request.getRequestId(), monitoredComponents);
+        }
+    }
+
+    private Set<SetMonitoringDatum.Type> convertCriteriaToMonitorType(List<MonitoringCriterium> criterion) {
+        Set<SetMonitoringDatum.Type> result = new HashSet<>();
```
Comment (akoshterek) [inline simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/GetMonitoringReportRequestHandler.java]: EnumSet is more appropriate for enums

**Comment by MounirKaroudEV** [comment]
Can i merge it now?

**Pull request merged**